Front-end refactoring for display on small screens

ewels opened the pull request:
HTML / CSS fixes to make the Bootstrap navbar component work nicely on small screens:

<img src="https://user-images.githubusercontent.com/465550/65338607-da11ff00-dbca-11e9-9d04-3367f722d6a7.png" width="500">

---

<img src="https://user-images.githubusercontent.com/465550/65338617-e0a07680-dbca-11e9-8556-a9a727d2aabe.png" width="500">

ewels commented:
Ok, I did some more stuff.. The workflow detail page is properly responsive now, with the content blocks shifting around to fit on varying screen widths:

<img src="https://user-images.githubusercontent.com/465550/65361649-9805af00-dc04-11e9-99a4-1496988ff9b7.png" height="500">

---

<img src="https://user-images.githubusercontent.com/465550/65361653-9a680900-dc04-11e9-9536-ad74f15cb1d0.png" height="500">

---

<img src="https://user-images.githubusercontent.com/465550/65361655-9cca6300-dc04-11e9-963e-9e838f86b682.png" height="500">

---

The other problem was that the sidebar didn't really work on mobile. So I made it into an off-canvas sliding sidebar (based on [this](https://github.com/BlackrockDigital/startbootstrap-simple-sidebar) example).

The sidebar is shown by default on big screens and hidden by default on small screens. There's a new button in the header to toggle it:

![image](https://user-images.githubusercontent.com/465550/65361695-d26f4c00-dc04-11e9-9455-851e4f8837fb.png)

To accomadate this new button I had to make a narrow version of the header image for tiny screens:

![image](https://user-images.githubusercontent.com/465550/65361703-e2872b80-dc04-11e9-8110-ec96111b95ce.png)

ewels commented on tower-web/src/app/modules/main/component/navbar/navbar.component.ts:
```diff
@@ -32,5 +32,12 @@ export class NavbarComponent implements OnInit {
+  toggleSidebar() {
+    console.log(this);
+    this.isSidebarToggled = !this.isSidebarToggled;
+    // TODO: Make this work properly instead of hacking with jQuery
```
Couldn't figure out how to do this nicely on the plane - just toggling a class on another element that's in a different scope? This works fine but is hacky..

ewels replied:
Looking at a couple of stackoverflow posts, it seems crazy complicated to do this with angular! Maybe my little one-liner jQuery hack isn't all that bad..?! 😆

ewels replied:
Hang on, I think this jQuery command stops the front end from building... 🤦‍♂ Have commented out in the last commit, so sidebar doesn't work now.

pditommaso replied:
Ouch! 😆

ewels replied:
Yes, I need to fix this 😅 Will let you know when it’s ready for testing / merging..

ewels replied:
Fixed! 🎉

ewels commented on tower-web/src/styles.scss:
```diff
@@ -26,42 +26,74 @@ code {
-  position: sticky;
-  top: 0;
-  flex: 0 0 25%;
+  // TODO: Make sticky-top behaviour work with navbar height
```
Currently this hides underneath the top-nav when you scroll, I need to fix that still.

ewels commented on tower-backend/src/main/resources/application.yml:
```diff
@@ -3,7 +3,7 @@ tower:
     contactEmail: "${TOWER_CONTACT_EMAIL:hello@tower.nf}"
     trustedEmails:
-      - '*@seqera.io'
+      - '*'
```
Not sure if you would rather leave this filled? I guess I should remove the bit in the readme about updating this if we leave it like this..

pditommaso replied:
I'm changing this so that any user is whited listed by default and optional `trustedEmails` can be provided to enable this feature.

ewels replied:
Updated PR to solve merge conflicts ✨

ewels commented:
![Sep-21-2019 00-41-48](https://user-images.githubusercontent.com/465550/65362873-67744400-dc09-11e9-84b4-588cf872981f.gif)

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/seqeralabs/nf-tower?pullRequest=144) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/seqeralabs/nf-tower?pullRequest=144) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/seqeralabs/nf-tower?pullRequest=144) it.</sub>

pditommaso commented:
> .. we ask that you sign our Contributor License Agreement before we can accept your contribution.

Hope you will appreciate you are not asked to print and sign a pdf any more 😆

ewels commented:
Hmm, I think that there's something not right about the CLA setup..?

![image](https://user-images.githubusercontent.com/465550/65393794-cfe83000-dd85-11e9-92d1-187e93459a33.png)

pditommaso commented:
It looks veeery slow, but waiting some seconds it shows it. 
<img width="1412" alt="Screenshot 2019-09-22 at 22 32 53" src="https://user-images.githubusercontent.com/816968/65394060-24d97580-dd89-11e9-9946-35337d1fd0dd.png">

The gist is [this](https://gist.github.com/pditommaso/68e9a605cfb17142a870217469490820) and basically it's the same CLA used by HashiCorp.

ewels commented:
Ok, got it in the end I think! Had several timeout errors and stuff, but on the third attempt (on desktop) it worked :)

Still need to fix the PR before merging though, sorry. Will try to have a look on Wednesday when I'm actually in work..! 😱

pditommaso commented:
No hurry! Thanks Phil.

ewels commented:
Ok, I'm getting close now.. I think I have everything displaying properly, and I am toggling the `this.isSideBarShown` variable in the `navbar` component (that's where the button is located):

https://github.com/seqeralabs/nf-tower/pull/144/files#diff-886265eb8786630007bb9ad80a052b35R40-R43

However, I'm not able to get the `home` component template to see this variable. I want to set the `toggled` class here:

https://github.com/seqeralabs/nf-tower/pull/144/files#diff-3a398b38f450e42a2e1ac4f3e062ab0eR8

Help would be welcome! 🆘

ewels commented:
I feel some kind of very simple `EventEmitter` like in [this SO post](https://stackoverflow.com/a/47924683/713980) must be the answer. But I can't get it to work... :/

ewels commented:
Ok, I got it to work! Via the use of an `EventEmitter`, following [this tutorial](https://angularfirebase.com/lessons/sharing-data-between-angular-components-four-methods/#Child-to-Parent-Sharing-Data-via-Output-and-EventEmitter).

Ok, I think this PR is now ready to review 🎉

ewels commented:
One thing: I wasn't 100% happy with my icon choice for the button that toggles the sidebar. I went for it because I thought it kind of resembled the list of workflow cards. But for this and any other style choices (or code, or anything really), I of course completely defer to you guys 🙇

pditommaso commented:
it looks the footer is broken 
<img width="1552" alt="Screenshot 2019-10-03 at 14 36 22" src="https://user-images.githubusercontent.com/816968/66127136-4dfcd000-e5eb-11e9-9e9e-321702ae8359.png">

pditommaso commented:
The collapsible sidebar is neat! Thanks a a lot for this, however there a couple of problems to solve:

* The first looks simple: it's missing the grey border on the right of the sidebar (see the arrow in the screenshot) 
* Then we should find a better place for the collapse/expand icon. On the left of the Tower logo does not work. What if the sidebar collapse to a thin vertical bar on left containing only a small expand icon? At some point, we could even use it to render quick run updates/info.

<img width="1552" alt="Screenshot 2019-10-03 at 14 40 56" src="https://user-images.githubusercontent.com/816968/66127823-e051a380-e5ec-11e9-9c4a-2d444c648fec.png">

ewels commented:
What’s wrong with the footer sorry?

Good spot with the grey border!

I did wonder if you’d say that about the new button - I didn’t love it either. I had initially thought of the vertical bar idea too before I started. The problem is that on mobile screens the thin vertical bar is either too thin to press easily or takes up a significant proportion of the screen. The button takes up no extra space and is easy to tap.

I’ll have a play with a vertical bar and see how it feels. If it is as I suspect above, how about having the button visible only on xsmall mobile phone screens? Then have the thin vertical bar on medium size up..

Phil

pditommaso commented:
Hey Phil 
 
> What’s wrong with the footer sorry?

The footer should be anchored to the bottom. Have a look at https://tower.nf/login

> The problem is that on mobile screens the thin vertical bar

Yeah, this is also true but Tower is a desktop-first app. Maybe we should have a version design specifically for the mobile, at the end it would be just a matter of having a separate html template. What do you think?

pditommaso commented:
Just linking as potential options to explore https://medium.com/better-programming/creating-angular-webapp-for-multiple-views-and-screen-sizes-50fe8a83c433

ewels commented:
No need for separate views here I think. It’s very easy to show or hide a few different elements at different screen sizes with CSS (and bootstrap). This is what I was suggesting above - your vertical bar for desktops and my navbar button for mobile. Best of both worlds 😁

pditommaso commented:
Deal 👍

ewels commented:
I started playing with different options, but then I thought.. Do we actually need the ability to collapse the sidebar on big screens at all?

How about on large screens, there is no change in behaviour. On small screens, with hide the sidebar by default and add the button as it is above.

ewels commented:
Ok, I think that this is a better compromise.. Let me know what you think!

pditommaso commented:
See your point, but IMO this would fragment even more the layout. I would prefer to keep the collapsible sidebar even on large screens.

ewels commented:
You're a hard man to please 😅 

Inspiration for potential UIs:
* [AdminLTE bootstrap dashboard theme](https://adminlte.io/themes/AdminLTE/index2.html)
* [Travis](https://travis-ci.com/) (simply shows active jobs at the bottom of the page)
* [Twitter](https://twitter.com/home) (icons + text on big screens, just icons on small page)
* [TweetDeck](https://tweetdeck.twitter.com/) (icons, collapse button at bottom)

Ok, I think my preferred implementation is like TweetDeck: to have a new button fixed at the bottom of the sidebar - `« Collapse`. Then in the collapsed state, it is wide enough to show coloured icons. These are the tick / cross / running icons for each workflow, plus a `»` to expand the sidebar again.

What do you think?

pditommaso commented:
> You're a hard man to please 

I'm really sorry .. but you opened the pandora box 😆 

> Ok, I think my preferred implementation is like TweetDeck

I agree. Adding also WorkMail screenshot that uses a similar design to get extra idea. 

<img width="1369" alt="Screenshot 2019-10-04 at 14 54 45" src="https://user-images.githubusercontent.com/816968/66209140-0774a780-e6b7-11e9-9a67-23a483950a0b.png">

<img width="1369" alt="Screenshot 2019-10-04 at 14 54 55" src="https://user-images.githubusercontent.com/816968/66209154-0c395b80-e6b7-11e9-9e52-4fef2c53cbd7.png">

If we go for the TweetDeck/WorkMail style what about the top navbar? I was imaging to keep it and just collapse the sidebar. Thought?

ewels commented:
Nice! Ok, I will have a play with refactoring that. It may take me a little while to find time again whilst I balance projects, but I'll do my best.

> If we go for the TweetDeck/WorkMail style what about the top navbar? I was imaging to keep it and just collapse the sidebar. Thought?

I was thinking we keep it..

What does WorkMail do on small screens? (ie. [Chrome Dev tools > Mobile preview](https://developers.google.com/web/tools/chrome-devtools/device-mode)).

pditommaso commented:
No hurry! Thanks for your patience! 

> I was thinking we keep it..

OK 

> What does WorkMail do on small screens?

Nothing ..

<img width="240" alt="Screenshot 2019-10-04 at 15 58 50" src="https://user-images.githubusercontent.com/816968/66213445-157af600-e6c0-11e9-8b2e-949445e2fd71.png">

pditommaso commented:
OK, let's focus this on fixing the UI on mobile and then the advanced sidebar will be implemented in a separate feature. Is this complete on your side?

ewels commented:
Hah, you should have never said _”no hurry”_! 😂

To be honest, separating out the sidebar stuff from the mobile stuff would take me as long as just finishing up the sidebar I think. I’ll try to find some time to come back to this ASAP.

pditommaso commented:
LOL. Anyhow, I'm saying that because we are planning to add more features and change the current behaviour, therefore it may overlap with what we were discussing here. Therefore I don't want to work on something we'll to change a lot in the near feature (unless it's something trivial!)

ewels commented:
I started playing with this again today. Will hopefully have a new version with you soon. At home now so can't send screenshots, but my current idea is to:

* Have a `Collapse` button alongside the search bar (half the width of the now-narrower search input)
* When collapsed, have small square buttons for expand and search (which will both expand the sidebar)  and if possible, also have icons for each workflow (running / failed / succeeded etc) which can be clicked on without expanding the sidebar.
    * On desktop, hovering over the icon will show a [tooltip](https://getbootstrap.com/docs/4.4/components/tooltips/) with the pipeline and run names.
* Revert behaviour so that the sidebar is always shown by default on page load until hidden
    * This would then be consistent between mobile and desktop.
    * Only remaining difference will be that on mobile the main page slides and on desktop the width changes

How does this sound?

----

I had a quick play implementing a narrow Amazon Mail vertical expand bar. One that kind of looked a bit like your first Amazon Mail example:


![image](https://user-images.githubusercontent.com/465550/69677150-95ab4f80-10a2-11ea-9904-44fce6bf4e91.png)

I really didn't like it on mobile though - to make it big enough to be tapped, it has to be fairly wide and it doesn't hold any information. So there's just a bit empty space which seems like a waste.

Looking at your [mobile screenshot](https://github.com/seqeralabs/nf-tower/pull/144#issuecomment-538409477) again now, I guess that the far left icons remain on mobile.

<img src="https://user-images.githubusercontent.com/465550/69677107-757b9080-10a2-11ea-82e4-e02188fc7296.png" height=300>

This makes sense, as they're wide enough to tap and semi-informative. This is pretty much the approach I'm aiming for with the above.

ewels commented:
Ok @pditommaso, now ready for review I think! New sidebar implemented:

![desktop-tower-sidebar](https://user-images.githubusercontent.com/465550/69734790-e74deb80-112f-11ea-9a78-accb4568c547.gif)


![mobile-tower-sidebar](https://user-images.githubusercontent.com/465550/69734811-ee74f980-112f-11ea-9f83-8147555e39d0.gif)

ewels commented:
Still todo:

- [ ] Fix merge conflicts
- [ ] Fix tooltips on the small sidebar buttons (I thought I had done this, but they don't work in the demo)
- [ ] Fix sticky sidebar scrolling on mobile

Hopefully this is enough for you to decide if the concept is acceptable though, these are relatively minor issues :)

pditommaso commented:
Wow, at first sight it looks amazing! will check soon in detail soon. 👏

ewels commented:
Great! I much prefer it. Also as it’s all contained in the sidebar the code is a lot simpler - I was able to get rid of all of the state emit / listening stuff.

pditommaso commented:
> Also as it’s all contained in the sidebar the code is a lot simpler - I was able to get rid of all of the state emit / listening stuff.

I love this ! :D

ewels commented:
Also, if I can figure out how to pass the currently active workflow ID as a variable to the sidebar workflow cards in order to have an `.active` class, I would like to highlight the card that corresponds to the active page...

pditommaso commented:
Impressive work Phil, I like it. I think in the future we may use collapse icons to access specific functions, but for now, it looks great. 

Just one thing: what about making the large collapse button just showing the arrow to leave more space to the search box? (I can do, I should be able doing that 😂)

pditommaso commented:
Also maybe the lens icons showed when the sidebar is collapsed could be removed, at the end it's just opening the sidebar as the arrow.

ewels commented:
> Just one thing: what about making the large collapse button just showing the arrow to leave more space to the search box?

Totally doable, but I personally prefer having things a bit more explicit if there's room. For example, even though I've been using tower for quite a long time now I only realised yesterday that you could collapse the error log on failed runs, and that was because I clicked it by accident 😅 (I changed that into a button that also says "collapse" in this PR also).

<img src="https://user-images.githubusercontent.com/465550/69833457-4dbc3200-1234-11ea-80e1-5d31b242647f.png" width=500>

How about as a compromise the button shrinks when the search input is focussed?

Phil

ewels commented:
![collapse-search-focus](https://user-images.githubusercontent.com/465550/69833675-adffa380-1235-11ea-8dfa-d3a29e887ec3.gif)

In fairness, could probably hide the collapse button entirely in this case too if we want.

ewels commented:
Ok, I think this is now ready for review.. I've fixed all of the outstanding issues I listed above and it seems to be working pretty well for me now on both mobile preview and desktop.

I refactored one or two bits I'd written and one or two bits I hadn't (eg. it's now `position:fixed` instead of `position: sticky` for the sidebar), plus I figured out how to highlight the active card (was actually pretty easy in the end!).

Let me know what you think 😊 

---

![sidebar-highlight-active](https://user-images.githubusercontent.com/465550/69835869-ece92580-1244-11ea-8fe6-90370315d938.gif)

pditommaso commented:
Great, thanks Phil! I agree with the collapse in the error but still not 100% sure about the one next to the search button. I think I'm going to remove the "collapse" text to keep it a bit simpler and merge everything. Great work!

pditommaso commented:
OK, I was too optimistic. The content in the Welcome/user/tokens pages is aligned to right instead of being centred. I guess you changed something in the main structured that altered them. Any clue?

pditommaso commented:
Also the footer is not rendered properly 

<img width="1428" alt="Screenshot 2019-11-29 at 16 27 46" src="https://user-images.githubusercontent.com/816968/69878435-3d07cc80-12c5-11ea-90ff-927ad5d83477.png">

ewels commented:
I’ll take a look 👍🏻

ewels commented:
Ok, both bugs fixed 👍 Let me know if you spot anything else that looks odd..

ewels commented:
Ok, and I've removed the `Collapse` text from the button now too. All done I think 👍 (unless you find more funny business)..

pditommaso commented:
Looks fixed! thanks!

pditommaso commented on tower-web/src/app/modules/main/component/workflow-status/workflow-status.component.ts:
```diff
@@ -25,11 +25,8 @@ export class WorkflowStatusComponent implements OnInit {
   constructor() {
     this.statusesTags = getAllTaskStatusesProgressStateTags();
-    this.statusesRows = chunk(this.statusesTags, 3)
   }
```
Quick question: what's the rationale of this change. I confess I don't know what `chunk` was doing here.

ewels replied:
`statusesRows` was how you were organising the coloured boxes into two rows of three:
![image](https://user-images.githubusercontent.com/465550/69974325-514a0480-1525-11ea-8a63-2f24429cbe21.png)

To make these fit whatever screen size is available, I refactored the front-end CSS so that these boxes stack into columns automatically. The same 2 rows x 3 columns on large screens, but also 3 x 2 on narrower screens and 6 x 1 on tiny screens. As I was able to do this only using CSS, I was able to delete the back-end code for `statusesRows` as it wasn't needed any more.

pditommaso commented on tower-web/src/assets/variables.scss:
```diff
@@ -11,12 +11,10 @@ $submitted-orange: #f57f27;
 
-$navbar-height: 64px;
+$navbar-height: 63px;
 $font-size-root: 0.90rem;
```
Why the navbar is 1px smaller? #curiosity :D

ewels replied:
I think I spotted a 1px gap between the sidebar and the nav..! That was it 🤷‍♂

pditommaso commented on tower-web/src/app/modules/main/component/workflow-main-tabs/workflow-main-tabs.component.html:
```diff
@@ -2,13 +2,13 @@
       <li class="nav-item">
-        <a class="nav-link" id="configuration-tab" data-toggle="tab" href="#configuration" role="tab" aria-controls="configuration">Configuration</a>
+        <a class="nav-link" id="configuration-tab" data-toggle="tab" href="#configuration" role="tab" aria-controls="configuration">Config<span class="d-none d-sm-inline">uration</span></a>
       </li>
```
wow, this is like a pro 🏆

ewels replied:
![image](https://user-images.githubusercontent.com/465550/69974626-eea53880-1525-11ea-86c7-71ea0fafa2c4.png)

pditommaso commented:
Ok, can't wait to release this! let's merge it! 🎉

ewels commented:
haha, yay!! 🚀 🎉

pditommaso commented:
OK, fixed to footer adding a nested `<div>` 07a5e68167ad93a6baf797838817be2f01ed367e. 

I've made also some other few changes [here](https://github.com/seqeralabs/nf-tower/commit/9a2b72e7d2615ed15c1b0df652107e7f8f88e4b2) and [here](https://github.com/seqeralabs/nf-tower/commit/f92005b2d3174d7691e97a3d50ead5c53b016035).

ewels commented:
Nice!
